fuse/file.go

```diff
@@ -3,103 +3,370 @@ package fuse
+	if !plugin.ReadAction().IsSupportedOn(f.entry) {
+		activity.Warnf(ctx, "FUSE: Non-contiguous writes (at %v) unsupported on %v", start, f)
+		return nil, fuse.ENOTSUP
+	}
+
+	return plugin.Read(ctx, f.entry, end-start, start)
+}
+
+var _ = fs.HandleFlusher(&file{})
+
+// Note that this implementation of Flush only calls plugin.Write if there were previous calls to
+// Write or Setattr. It doesn't check whether the data that's there matches what we're writing.
+func (f *file) Flush(ctx context.Context, req *fuse.FlushRequest) error {
+	f.mux.Lock()
+	defer f.mux.Unlock()
+	activity.Record(ctx, "FUSE: Flush %v: %+v", f, *req)
+
+	if _, ok := f.writers[req.Handle]; !ok {
+		return nil
+	}
+
+	// If this handle had an open writer, write current data.
+	dataLen := int64(len(f.data))
+	if f.isFileLike() {
+		// Only file-like entries keep data and readSize in sync.
+		if uint64(dataLen) > f.readSize {
+			panic("Size was not kept up-to-date with changes to data.")
+		}
+
+		if uint64(dataLen) < f.readSize {
+			// Missing some data, load the remainder before writing.
+			data, err := f.load(ctx, dataLen, int64(f.readSize))
+			if err != nil && err != io.EOF {
+				return err
+			}
+			f.data = append(f.data, data...)
+
+			// If still too small then something increased the size beyond the original.
+			// Fill with null characters.
+			if sz := uint64(len(f.data)); sz < f.readSize {
+				f.data = append(f.data, make([]byte, f.readSize-sz)...)
+			}
+		}
+	}
+
+	if err := plugin.WriteWithAnalytics(ctx, f.entry.(plugin.Writable), f.data); err != nil {
+		activity.Warnf(ctx, "FUSE: Error writing %v: %v", f, err)
+		return err
 	}
-	activity.Record(ctx, "FUSE: Read %v/%v bytes starting at %v from %v: %v", len(data), req.Size, req.Offset, fh.id, err)
-	resp.Data = data
-	return err
+
+	// Non-file-like entries start from scratch on each Write operation, and have their cache
+	// invalidated whenever we write to them because we can't accurately model their readable data.
+	if !f.isFileLike() {
+		f.releaseWriter(req.Handle)
+	}
+	return nil
 }
 
-func (fh fileHandle) Write(ctx context.Context, req *fuse.WriteRequest, resp *fuse.WriteResponse) error {
-	n, err := plugin.WriteWithAnalytics(ctx, fh.w, req.Offset, req.Data)
-	resp.Size = n
-	activity.Record(ctx, "FUSE: Write %v/%v bytes starting at %v from %v: %v", n, len(req.Data), req.Offset, fh.id, err)
-	return err
+var _ = fs.NodeSetattrer(&file{})
+
+func (f *file) Setattr(ctx context.Context, req *fuse.SetattrRequest, resp *fuse.SetattrResponse) error {
+	f.mux.Lock()
+	defer f.mux.Unlock()
+	activity.Record(ctx, "FUSE: Setattr[%v] %v: %+v", req.Handle, f, *req)
+
+	if req.Valid.Size() {
+		if !req.Valid.Handle() {
+			// No guarantee we'll ever write the change. If this is ever necessary, we could update it
+			// to immediately do a plugin.Write.
+			return fuse.ENOTSUP
+		}
+
+		// Ensure handle is in list of writers.
+		f.writers[req.Handle] = struct{}{}
```

Comment:
Might be helpful to mention somewhere that this will likely be (only) called to truncate the file for writes. That helps clarify why we're including the handle in the list of writers.

Reply:
I'll elaborate in the comment. It's not only used to truncate the file (to size 0), you could directly call `truncate` on it and supply any size.